src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs

```diff
@@ -0,0 +1,3144 @@
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestNoInvocationAfterIfbranch()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> h, int x)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        if (x == 1)
+        {
+            [|i|].Union([|h|]).ToList();
+            [|i|].Join([|h|], n => n, n => n, (n, m) => n + m).ToList();
+        }
+    }
+}";
+            await VerifyCS.VerifyAnalyzerAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(h As IEnumerable(Of Integer), x As Integer)
+            Dim i = Enumerable.Range(1, 10)
+            If x = 1 Then
+                [|i|].Union([|h|]).ToList()
+                [|i|].Join([|h|], Function(n) n, Function(n) n, Function(m, n) m + n).ToList()
+            End If
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestOneInvocationAfterIfBranch1()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(int b, int[] c)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        if (b == 1)
+        {
+            [|i|].SingleOrDefault();
+            c.Max();
+        }
+        else if (b == 3)
+        {
+            [|i|].ToList();
+            c.Min();
+        }
+        else
+        {
+            [|i|].Min();
+            c.Average();
+        }
+
+        [|i|].Sum();
+        c.First();
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(b As Integer, c As Integer())
+            Dim i = Enumerable.Range(1, 10)
+            If b = 1 Then
+                [|i|].SingleOrDefault()
+                c.Max()
+            Else If b = 3
+                [|i|].ToList()
+                c.Min()
+            Else
+                [|i|].Min()
+                c.Average()
+            End If
+
+            [|i|].Sum()
+            c.First()
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestOneInvocationAfterIfBranch2()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(int b, IEnumerable<int> j)
+    {
+        var i = Enumerable.Range(1, 10).ToArray();
+        i.First();
+        [|j|].SingleOrDefault();
+
+        if (b == 1)
+        {
+            i.First();
+            [|j|].First();
```

Comment:
Isn't it only return the `First` element in the sequesnce? I would not warn for the parameterless overload of `First()`

Reply:
That's interesting. Then do we also want to exclude `ElementAt(0)` ?

Reply:
Maybe, in case it can be excluded easily

Reply:
I have been thinking about this for some time when editing the code.
I think this might be two kinds of  'multiple enumerations'.
1. Multiple call to the `MoveNext()` mthod, like `i.ElementAt(10)`. Calling this over `IEnumerable<T>` is the same as doing an indexing operation over a linked list.

2. The source of `IEnuemrable<T>` itself might change.
```
void Bar()
{
    IEnumerable<int> result = FetchDataFromDatabase();
    var a = result.First();
    WriteSomeDataIntoDataBase();
    var b = result.First();
}
```
In this case, even it is just called `First`, I still feel we should flag it. Because `a` and `b` might be two different values. And `First` is really querying the DB twice.

My purpose solution here would be, for a given variable which is `IEnumerable<T>`, if it is assigned from `System.Linq.Range()` or `System.Line.Repeat()`, we can safely ignore the usage of `First`, `FirstOrDefault`, etc...
Ex:
```
void Bar()
{
     var c = Enumerable.Range(1, 10000);
     c.First();
     c.FirstOrDefault();
}
```
No diagnostic should be reported.

But if the variable is assigned from some unknown method call, we should consider `First`, `FirstOrDefault`, etc...

Reply:
> In this case, even it is just called First, I still feel we should flag it. Because a and b might be two different values.
> And First is really querying the DB twice.

Did not understood, for me `Because a and b might be two different values.` is the reason for not flagging it, becaues the values could be different its NEED to query twice

Reply:
My point is this is a good case that people might misuse the Linq. (The first impression for this is that `a` should equal to `b`, but in fact because it queries DB twice, the value would be different)

Reply:
I am not sure how that could be a misuse of Linq, anyway i don't think we need to evaluate subsequent calls of `First()` or try to undestand the user expectations for those calls, because that doesn't seem to relate to the purpose of this analyzer (detecting multiple Enumerations), `First()` would never call `MoveNext()` more than once (by the point 1 you mentioned)

Reply:
OK I agree we don't add this behavior in the first revision of this analyzer.
Probably we want to wait for some feedback to see if customers really want this. 
I do really feel this might lead to bugs in reality. but we can wait to see it

Reply:
I am going to talk about this with our PM, because in the original post https://github.com/dotnet/roslyn-analyzers/issues/3946#issue-672955039
I think the post is showing 

> Multiple call to the MoveNext() mthod, like i.ElementAt(10). Calling this over IEnumerable<T> is the same as doing an indexing operation over a linked list.

But I want to double check if the customer may/may not want to 

> The source of IEnuemrable<T> itself might change.`

If needed I can bring that to the runtime design meeting again.

Reply:
I thnk you are mixing up my comment with other methods (like `ElementAt(10)`) My comment here is only about `First()` which would never make multiple call to the `MoveNext()`. For `ElementAt(10)` it would make multiple  `MoveNext()` therefore you should check/flag multiple calls

Reply:
No, I understand your idea here. `First()` or `FirstOrDefault()` will never call 'MoveNext()' twice. And for method like `ElementAt()` this is a different thing.

My question is mainly regarding should the analyzer report diagnostic when calling `First` or `FirstOrDefault` (the second example I posted above) This is not put in the initial post, but I want to verify with the PM about this

Reply:
> but I want to verify with the PM about this

Sure thing

Reply:
Based on the offline email discussion, I will mark this as resovle.
But thank you to raise this question. : )